### ISSUE TITLE
eliminate shared mutable state from reactor implementations

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2811]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#2811|
|Type|enhancement|
|Created|2007-09-04 01:22:10Z|

Most of the reactors included use module-level dictionaries to track their internal state.  This has some drawbacks:

  * limits them to one instance, effectively
  * makes testing messy
  * doesn't read nicely

This was probably done for the early reactors in support of an optimization which likely is no longer actually an optimization.  For later reactors, the pattern was simply copied.

Each reactor should keep all of its state as instance attributes.  One instance should have no effect on others.

<details><summary>Searchable metadata</summary>

```
trac-id__2811 2811
type__enhancement enhancement
reporter__exarkun exarkun
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1188868930000000 1188868930000000
changetime__1188998055000000 1188998055000000
version__None None
owner__ 
cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun removed owner|
|-|-|

Done in isolated-state-2811

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve set owner to @exarkun|
|-|-|

 * win32eventreactor didn't get the same treatment
 * unused import Failure in epollreactor
 * there are a few has_key remaining in selectreactor and pollreactor, it would be good to remove them
 * it looks like _reads and _writes could be defined in posixbase, but I realized that sometimes they hold ```FileDescriptor``` object, sometimes integer file descriptor, that's not really consistent. They probably should be sets too, but both points are not relevant here.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun removed owner|
|-|-|

Fixed the first three points, left the fourth alone.  After this is merged we should be able to more reasonably write _unit_ tests for all this code, and then I'll be more comfortable with that kind of refactoring.

I haven't run the tests with win32eventreactor.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve set owner to @exarkun|
|-|-|

 * there is a huge typo in gtkreactor :/ You used ```self.__simtag``` instead of ```self._simtag```. That makes me very uncomfortable about the problems I didn't spot.
 * the "grumble" in gtk reactor has to do with the 1010 multiplication I think
 * epoll copyright should be updated.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to @therve|
|-|-|

Luckily the test suite completely explodes if you actually try to use PortableGtkReactor with that bug in place.  Unluckily no buildslave is configured to run that reactor.

Fixed it, and the epollreactor copyright.  Should I do something about the grumble comment?

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve set owner to @exarkun|
|-|-|

I guess it's ok, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set status to `closed`|
|-|-|

(In [21216]) Merge isolated-state-2811

Author: exarkun Reviewer: therve Fixes [#2188](https://github.com/twisted/twisted/issues/2188) Refs [#2545](https://github.com/twisted/twisted/issues/2545)

Remove global state from reactor implementations in favor of per-instance state.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
